gwos/contracts/state-validator/src/verifications/submit_block.rs

```diff
@@ -690,58 +732,155 @@ fn check_block_withdrawals(block: &L2BlockReader) -> Result<(), Error> {
     Ok(())
 }
 
+// Assert l2block.timestamp == post_global_state.tip_block_timestamp
+// Assert l2block.timestamp <= l1tx.since
+// Assert l2block.timestamp <=
+//      post_global_state.last_finalized_timepoint
+//      + rollup_config.finality_as_duration()
+//      + 4h
+// Assert l2block.timestamp >=
+//      post_global_state.last_finalized_timepoint
+//      + rollup_config.finality_as_duration()
+//      - 4h
 fn check_block_timestamp(
-    prev_global_state: &GlobalState,
+    rollup_config: &RollupConfig,
     post_global_state: &GlobalState,
     block_timestamp: u64,
 ) -> Result<(), Error> {
-    let prev_version: u8 = prev_global_state.version().into();
-    let post_version: u8 = post_global_state.version().into();
-
-    if 0 == post_version && post_global_state.tip_block_timestamp().unpack() != 0 {
-        debug!("v0 global state tip block timestamp isn't 0");
-        return Err(Error::InvalidPostGlobalState);
+    let post_block_timestamp: u64 = post_global_state.tip_block_timestamp().unpack();
+    if block_timestamp != post_block_timestamp {
+        debug!(
+            "[check_block_timestamp] block.timestamp is not same as post_global_state.tip_block_timestamp, {} != {}",
+            block_timestamp, post_block_timestamp
+        );
+        return Err(Error::InvalidBlock);
     }
 
-    // NOTE: Downgrade already checked in main
-    if 0 == post_version {
-        debug!("[check block timestamp] skip block timestamp");
-        return Ok(());
+    let post_version: u8 = post_global_state.version().into();
+    if post_version >= 1 {
+        let rollup_input_since = Since::new(load_input_since(0, Source::GroupInput)?);
+        let rollup_input_timestamp = match (
+            rollup_input_since.is_absolute(),
+            rollup_input_since.extract_lock_value(),
+        ) {
+            (true, Some(LockValue::Timestamp(time_ms))) => time_ms,
+            _ => return Err(Error::InvalidSince),
+        };
+        if rollup_input_timestamp <= block_timestamp {
+            debug!(
+                "bilibili rollup_input_timestamp <= block_timestamp"
+            );
+            return Err(Error::InvalidPostGlobalState);
+        }
     }
 
-    let rollup_input_since = Since::new(load_input_since(0, Source::GroupInput)?);
-    if !rollup_input_since.is_absolute() {
-        return Err(Error::InvalidSince);
+    if post_version >= 2 {
```

Comment:
Use `else if`. We should skip the previous check designed for version 1.

Reply:
I think it should not be `else if`, the above check is for versions that `version >= 1` but not only for version 1.

Reply:
You are right. This feature can be disabled at v2.
Another new commit  ba1921392e65a5848cd17da29bc626d8b3d8a0f2 do it.